cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkUtils.java

```diff
@@ -185,6 +203,65 @@ public static String prepareSparkResources(File tempDir, Map<String, LocalizeRes
     return sparkAssemblyJar.getName();
   }
 
+  /**
+   * Returns the Spark environment setup via the start up script.
+   */
+  public static synchronized Map<String, String> getSparkEnv() {
+    if (sparkEnv != null) {
+      return sparkEnv;
+    }
+
+    Map<String, String> env = new LinkedHashMap<>();
+    for (Map.Entry<String, String> entry : System.getenv().entrySet()) {
+      if (entry.getKey().startsWith(SPARK_ENV_PREFIX)) {
+        env.put(entry.getKey().substring(SPARK_ENV_PREFIX.length()), entry.getValue());
+      }
+    }
+
+    // Spark using YARN and it is needed for both Workflow and Spark runner. We need to set it
+    // because inside Spark code, it will set and unset the SPARK_YARN_MODE system properties, causing
+    // fork in distributed mode not working. Setting it in the environment, which Spark uses for defaults,
+    // so it can't be unset by Spark
+    env.put("SPARK_YARN_MODE", "true");
+
+    sparkEnv = Collections.unmodifiableMap(env);
+    return sparkEnv;
+  }
+
+  /**
+   * Returns the environment for the Spark client container.
+   */
+  public static Map<String, String> getSparkClientEnv() {
+    Map<String, String> env = new LinkedHashMap<>(getSparkEnv());
+
+    // The spark-defaults.conf will be localized to container
+    // and we shouldn't have SPARK_HOME set
+    env.put(SPARK_CONF_DIR, "$PWD");
+    env.remove(SPARK_HOME);
+
+    // Spark using YARN and it is needed for both Workflow and Spark runner. We need to set it
+    // because inside Spark code, it will set and unset the SPARK_YARN_MODE system properties, causing
+    // fork in distributed mode not working. Setting it in the environment, which Spark uses for defaults,
+    // so it can't be unset by Spark
+    env.put("SPARK_YARN_MODE", "true");
+
+    return Collections.unmodifiableMap(env);
+  }
+
+  @Nullable
+  public static File locateSparkDefaultsConfFile(Map<String, String> env) {
+    File confFile = null;
+    if (env.containsKey(SPARK_CONF_DIR)) {
+      // If SPARK_CONF_DIR is defined, then the default conf should be under it
+      confFile = new File(env.get(SPARK_CONF_DIR), SPARK_DEFAULTS_CONF);
+    } else if (env.containsKey(SPARK_HOME)) {
+      // Otherwise, it should be under SPARK_HOME/conf
+      confFile = new File(new File(env.get(SPARK_HOME), "conf"), SPARK_DEFAULTS_CONF);
+    }
+
+    return confFile == null || !confFile.isFile() ? null : confFile;
```

Comment:
Is it possible to have a setup where `spark-default.xml` is not present on the cluster? If not then should we throw exception here to fail the spark program?